---
title: "Ralstonia H16 genome re-annotation"
output: html_notebook
---

<!-- Link [R Markdown](http://rmarkdown.rstudio.com) -->

# An attempt to re-annotate COG groups

The _Ralstonia eutropha_ H16 genome was annotated using the EMBL-maintained tool 
[eggNOG](http://eggnogdb.embl.de/#/app/home). It annotates all genes/proteins 
based on sequence similarity and groups them into homology groups (Cluster of 
Orthologous Groups, COG). Other annotations used for _Ralstonia_ are mainly 
derived from _uniprot_ or _KEGG_. It is important to mention that KEGG Brite 
annotation is not unambiguous like COG, but one gene can be mapped to multiple 
functions/pathways. This behavior is more realistic, but here, KEGG annotation 
was flattened to the most frequent pathway annotation per gene only. Other sources 
used for re-assignment of genes to groups were _eggNOG_ and  _uniprot_ functional 
annotation and mainly the following sources from literature:

- Cramm R, _Genomic view of energy metabolism in Ralstonia eutropha H16_,
  Journal of Molecular Microbiology and Biotechnology, 2008.
- Lu et al., _Ralstonia eutropha H16 as a Platform for the Production of Biofuels,_
  _Biodegradable Plastics, and Fine Chemicals from Diverse Carbon Resources_, 
  Biotechnology for Biofuel Production and Optimization, 2016.

### Problem definition 

A massive problem with the original COG annotation based on eggNOG prediction
is that proteins are assigned to apparently 'wrong' groups.
Grouping/clustering is always subjective and never perfect, however, highly
functionally related proteins should end up in similar groups. The prime 
example are the two Rubisco subunits, cbbS and ccbL, that end up in different
groups, Energy production and Coenzyme transport. This script aims to
mitigate this problem by a set of re-organizing steps:

1. Reassign obviously falsely classified proteins to other existing groups
   Example: move cbbL from Coenzyme to other group
   
2. Divide large groups into sub-groups that are more meaningful for Ralstonia
   Example: Energy production into carbon-fixation/1-carbon metabolism
   and electron transport (hydrogenases, formate-dehydrogenases, ...)
   
3. Investigate candidates in Function unknown/NA and assign to groups

### Task 1: move falsely annotated proteins

Load required packages and original genome annotation.
The original _Ralstonia eutropha_ database is based on the 
[uniprot reference genome](https://www.uniprot.org/uniprot/?query=proteome:UP000008210) 
of _Ralstonia eutropha_ H16 a.k.a. _Cupriavidus necator_ H16.

```{r}
# packages
library(tidyverse)
library(lattice)
library(latticeExtra)
library(Rtools)

# data frame
df <- read_csv("Ralstonia_H16_genome_annotation.csv")
```

One of the most striking problems is mis-annotation of Calvin cycle proteins (_cbb_ operon).
We can have a look the distribution of COG terms for the _cbb_ operons. There are two _cbb_
operons, one on the plasmid ('PHGxxxx') and one on chromosome 2 ('H16_Bxxxx') with
around 12 genes each.

```{r}
# overview of cbb genes
df %>% filter(grepl("cbb", gene_name)) %>% select(gene_name, COG_Process)

# and plot
df %>% filter(grepl("cbb", gene_name)) %>% pull(COG_Process) %>% 
   table %>% enframe %>% arrange(desc(value)) %>%
   barchart(value ~ substr(name, 1, 16) %>% factor(., unique(.)), ., horizontal = FALSE,
      par.settings = custom.lattice, border = NA,
      scales = list(x = list(rot = 30))
   )

```

Several genes/proteins of the Calvin cycle functionally overlap with other pathways.
For example, there are 2 copies of FBP (fructose-1,6-bisphosphatase) and 
fructose-bisphosphate aldolase in the _cbb_ operons. Both enzymes are also important in
gluconeogenesis and therefore another branch of carbohydrate metabolism. 
However, the _R. eutropha_ genome contains more copies of these enzymes (iso-enzymes?) 
that can carry out the reaction.

The same is true for enzymes of the Calvin cycle that share pathway membership
with the reductive branch of the pentose phosphate pathway (PPP), like transketolase (tkt).
Also those have counterparts/iso-enzymes at other genomic locations. On the risk of
not judging each single protein correctly, we can allocate those to existing COG group
'Carbohydrate transport and metabolism'

```{r}
# reassign cbb genes to carbohydrate metabolism, following KEGG;
# explicitly exclude regulator cbbR

df_new <- df %>% 

   mutate_cond(
      grepl("cbb[^R]", gene_name), 
      COG_ID = "G",
      COG_System = "Metabolism",
      COG_Process = "Carbohydrate transport and metabolism"
   )
```

Other groups with falsely assigned proteins seem to be "Post-translational modification", 
"Coenzyme transport and metabolism", and "Nucleotide transport and metabolism". 
These groups comprise a particulary diverse set of genes which becomes clear from manual inspection. 
To be more systematic, we can look at the KEGG pathway assignments for the genes in
each of these categories.


```{r}
# filter and summarize table by counting genes per KEGG pathway
df_new %>% filter(grepl("Coenz|Nucleo|Post", COG_Process)) %>% 
   group_by(COG_Process, Pathway) %>% 
   summarize(n_genes = length(Pathway)) %>%
   arrange(COG_Process, desc(n_genes)) %>%
   filter(Pathway != "Hypothetical") %>%
   
   # and plot result
   barchart(n_genes ~ Pathway %>% factor(., unique(.))| COG_Process, ., 
      horizontal = FALSE, border = NA, par.settings = custom.lattice, 
      xlab = "KEGG pathway [ordered by no of member]", layout = c(1, 3),
      scales = list(alternating = FALSE, x = list(at = 1:10*10))
   )

```

We can see that there is relatively little overlap between pathways of the three
selected, broader COG groups. That's good because it means that COG categories
contain distinct groups of genes. A closer look nevertheless reveals that some of the genes
in 'Coenzyme' group clearly belong to other groups ('pyruvate metabolism' 
matches better to 'carbohydrate', 'ribosome biogenesis' to 'translation', and so on).

We can construct rules to re-assign genes to COG groups by matching corresponding 
(fine-grained) KEGG pathways to the (coarse) COG counterparts in cases that are clear.
It is also reasonable to re-assign all genes related to PHA/PHB metabolism to the
'Lipds transport and metabolism' group.


```{r}
# string together a series of conditional mutations
#
df_new <- df_new %>%
   
   # -- ribosome to translation --
   mutate_cond(
      grepl("ribosome", Pathway), 
      COG_ID = "J",
      COG_System = "Information storage and processing",
      COG_Process = "Translation, ribosomal structure and biogenesis"
   ) %>%
   
   # -- pyruvate/glycolysis to carbohydrate --
   mutate_cond(
      grepl("Pyruvate|Glycolysis", Pathway) & grepl("Coenz|Nucleo", COG_Process), 
      COG_ID = "G",
      COG_System = "Metabolism",
      COG_Process = "Carbohydrate transport and metabolism"
   ) %>%
   
   # -- tRNA synthesis to translation --
   mutate_cond(
      grepl("Transfer RNA", Pathway) & grepl("Function unknown", COG_Process), 
      COG_ID = "J",
      COG_System = "Information storage and processing",
      COG_Process = "Translation, ribosomal structure and biogenesis"
   ) %>%
   
   # -- TCA to energy --
   mutate_cond(
      grepl("TCA", Pathway), 
      COG_ID = "C",
      COG_System = "Metabolism",
      COG_Process = "Energy production and conversion"
   ) %>%
   
   # -- PHA/PHB genes to lipids --
   mutate_cond(
      grepl("[Pp]ha[ABCPYZ]", gene_name), 
      COG_ID = "I",
      COG_System = "Metabolism",
      COG_Process = "Lipid transport and metabolism"
   )

```



### Task 2: Divide large groups into sub-groups

The prime example is the large COG group 'Energy production' which contains a lot 
and very different metabolic pathways. This could for example be divided into TCA cycle, 
electron transport chain/oxidative phosphorylation/ATP synthase, and 
H2-hydrogenases/formate-dehydrogenases/carbon monoxide dehydrogenases (autotrophy).
The new COG categories also get a unique one-letter identifier (```COG_ID```). 

There is functional overlap between some of these assignments, so that they should be taken with
a pinch of salt. For example, Succinate dehydrogenase (Cytochrome b subunit) (_sdhC_, _H16_A263_)
is a membrane protein catalyzing oxidation of succinate and transfer of electrons to
a universal quinone acceptor. It thus plays a role in oxidative phosphorylation and 
electron transport, but also in the TCA cycle.


```{r}
# add some finer-grained 'COG' groups
df_new <- df_new %>% 
   
   # -- new group oxidative phosphorylation --
   mutate_cond(
      grepl("Oxidative phosphorylation", Pathway) | 
         grepl("nuo|nar[GHIJ]|nir|nor|nos|H16_B2315|H16_A0344|H16_A0346", gene_name) |
         grepl("[Cc]ytochr|quinol oxidase|[Qq]uino.*(oxido)?reductase|NADH.*(flavin|dehydro|reduct)", protein_name) & 
         !grepl("Post-trans", COG_Process), 
      COG_ID = "X",
      COG_System = "Metabolism",
      COG_Process = "Oxidative phosphorylation"
   ) %>%
   
   # -- new group autotrophic energy generation --
   mutate_cond(
      grepl("hox|fdh|fdo|fds|fdw", gene_name) |
         grepl("hup[SL]|hya|hyf", eggNOG_name) | 
         grepl("[Ff]ormate (de)?hydro|[Cc]arbon monox", protein_name), 
      COG_ID = "A",
      COG_System = "Metabolism",
      COG_Process = "Autotrophic energy generation"
   )

```

There is one group, ```Chromatin structure and dynamics```, that has extremely
few member genes as it is apparently important in eukaryotic cells but not 
prokaryotic ones. We will therefore merge it with group ```Cell cycle control, cell division, ...```.

```{r}
df_new <- df_new %>% 
   
   mutate_cond(grepl("Chromatin struct", COG_Process), 
      COG_ID = "D",
      COG_System = "Cellular processes and signaling",
      COG_Process = "Cell cycle control, cell division, chromosome partitioning"
   )

```


### Task 3: Assign new groups to 'Function unknown'/'NA' annotated genes

The final step is to take a look at the COG category 'Function unknown' or 'NA'.
It's a large group where the eggNOG algorithm failed to assign a COG category,
but other resources are available and can be used to pull out known genes and 
assign them to COG functional categories.

_Ralstonia_-specific candidates could be hyp/hox genes, nitrate respiration genes,
special transporters, PHA and other storage compounds, and many more. While some 
of the genes lack a COG annotation, they are annotated by more specific KEGG pathways 
or GO terms which can be used to assign them to COG groups.

```{r}
# let's have a look on KEGG Pathways within genes annotated as 'Function unknown'
df_new %>% 
   
   # summarize KEGG Pathways again
   filter(COG_Process == "Function unknown") %>%
   group_by(Pathway) %>%
   summarize(n_genes = length(Pathway)) %>%
   arrange(desc(n_genes)) %>%
   
   # remove 'hypothetical' and low-abundant pathways
   filter(Pathway != "Hypothetical" & Pathway != "Function unknown", n_genes >= 5) %>% 
   #pull(n_genes) %>% sum
   
   # and plot result
   barchart(n_genes ~ Pathway %>% factor(., unique(.)), ., 
      horizontal = FALSE, border = NA, par.settings = custom.lattice, 
      xlab = "KEGG pathway (>= 5 genes)", 
      scales = list(alternating = FALSE, x = list(rot = 30))
   )

```

Around 300 genes have KEGG pathway annotation, and by far the most of those 
are related to environmental interaction: transporters, sectretion systems, 
quorum sensing, chemotaxis, defense. These can be assigned to the corresponding 
COG groups. Again, there are several simplifications in place. Not all 'transport'
related genes are in charge of inorganic ion transport yet are assigned to this 
group. The transported substrate is often not known, and could among others 
comprise inorganic ions, small metabolites, toxins and xenobiotics, signaling 
molecules, and peptides.

```{r}
df_new <- df_new %>% 
   
   # -- move transport-related genes --
   mutate_cond(
      COG_Process == "Function unknown" & 
         grepl("[Tt]ransport", Pathway) | 
         grepl("[Tt]ransport", eggNOG_description), 
      COG_ID = "P",
      COG_System = "Metabolism",
      COG_Process = "Inorganic ion transport and metabolism"
   ) %>%
   
   # -- move secretion-related genes --
   mutate_cond(
      COG_Process == "Function unknown" & 
         grepl("[Ss]ecret", Pathway) | 
         grepl("[Ss]ecret", eggNOG_description), 
      COG_ID = "U",
      COG_System = "Cellular processes and signaling",
      COG_Process = "Intracellular trafficking, secretion, and vesicular transport"
   ) %>%
   
   # -- move genes related to defense/TA systems/cell cycle/division --
   mutate_cond(grepl("Defense", COG_Process) | (
         COG_Process == "Function unknown" & 
         grepl("defense|Quorum|Chromoso", Pathway) | 
         grepl("toxi[cn]", eggNOG_description)), 
      COG_ID = "D",
      COG_System = "Cellular processes and signaling",
      COG_Process = "Cell cycle control, cell division, chromosome partitioning"
   )
```

There are still many KEGG categories annotated with realtively few members. 
For those we can devise a more automated procedure to assisgn COG groups. We use 
a function that selects that most abundant COG terms for a given KEGG pathway 
and assign this term instead of "Function unknown". There is a risk of false 
assignment but the number and mass fraction of these proteins is low, compared to
the known and well annotated proteome.

```{r, results = 'hide'}
# function to fill in missing COG terms with the most abundant COG term within the 
# KEGG Pathway. In other words fill in COG terms derived from similar neighbors.
replace_NA_terms <- function(x) {
   # determine most abundant COG terms annotated for pathway
   terms = table(x) %>% sort(decreasing = TRUE); 
   terms %>% set_names(substr(names(.), 1, 10)) %>% print
   replace(x, x == "Function unknown", names(terms)[1])
}

# execute the function on annotation data frame
df_new <- df_new %>%
   
   # group by Pathway
   group_by(Pathway) %>%
  
   # exclude meaningless KEGG pathways
   mutate_cond(!Pathway %in% c("Function unknown", "General function prediction only", "Hypothetical"), 
      COG_Process = replace_NA_terms(COG_Process)
   )
  
```

### Summary and evaluation

Finally the changes that were made in COG annotation have to be summarized and 
compared to the original annotation obtained from eggNOG. We combine the two 
```data.frame```s and plot them as comparison of genes per COG group. We
can also calculate the change in gene number per group, and finally the resulting
new table is saved as ```*.csv```.

```{r}
# construct a summary table
df_summary <- bind_rows(
   df %>% group_by(COG_Process) %>% 
      summarize(n_genes = length(COG_ID)) %>% 
      mutate(type = "original"),
   df_new %>% group_by(COG_Process) %>% 
      summarize(n_genes = length(COG_ID)) %>% 
      mutate(type = "improved")
)

# plot improved versus original COG classification
barchart(n_genes ~ COG_Process %>% substr(1, 25) %>% factor(., unique(.)), 
   df_summary %>% arrange(desc(n_genes)), groups = type, origin = 0,
   horizontal = FALSE, border = NA, par.settings = custom.lattice, 
   xlab = "COG Process", ylab = "n_genes",
   scales = list(alternating = FALSE, x = list(rot = 30)),
   panel = function(x, y, ... ){
      panel.grid(h = -1, v = -1, col = grey(0.9))
      panel.barchart(x, y, ...)
      panel.key(c("re-assigned", "original"), pch = 15)
   }
)

# how many genes have changed?
df_summary %>% spread(type, n_genes) %>% 
   mutate_at(vars(2:3), function(x) replace_na(x, 0)) %>%
   mutate(changed = improved - original) %>%
   select(COG_Process, changed)

# save the revised table
write_csv(df_new, "Ralstonia_H16_genome_re_annotation.csv")

```






